File: backend/packages/billing/resume.py
"""Resuming a run the plan's window paused.

A run that exhausts the plan's quota does not fail — it PAUSES
(``guard.QuotaExhausted`` → ``video_projects.status = "paused_quota"``), keeping
everything it produced. What was missing until now was the other half: starting
again re-ran the whole pipeline, so the user paid a second time for work they
already had.

This module holds the small, shared pieces of the fix:

* the **resume ticket** (``video_projects.resume_state``) — written when a paid
  stage starts, so the SERVER knows which stage is in flight and exactly how to
  redo it; cleared when that stage succeeds. A paused project therefore carries
  the worker task + kwargs to re-enqueue, the estimate kind that prices ONLY
  the remaining stage, and the window/reset the pause hit;
* ``estimate_for`` — the remaining work's estimate, rebuilt from the ticket
  (never from the whole project): resuming charges for the interrupted stage,
  never for the boundaries already behind it;
* ``quota_view`` — whether the plan's windows still cover that estimate, when
  they roll, and what the top-up balance would have to pay if they do not.
  Percentages and baht only; never a token count (docs/token-billing-plan.md §2).

The endpoints that use all this are ``GET``/``POST /videos/{uid}/resume`` in
services/api/routers/videos_local.py, which also documents the client contract.
"""

from __future__ import annotations

from datetime import UTC, datetime
from typing import Any

from sqlalchemy.ext.asyncio import AsyncSession

from packages.billing import estimate as estimator
from packages.billing import limits as limits_mod
from packages.billing import runs, wallet
from packages.billing.accounts import get_account
from packages.billing.estimate import Estimate
from packages.db.models.core_auth import User

#: Bumped when the ticket's shape changes in a way a reader must notice. A
#: ticket from an older version is ignored rather than guessed at — the resume
#: endpoint then falls back to inferring the stage from the files on disk.
TICKET_VERSION = 1

#: Stages the SERVER runs and charges for. Everything else in the pipeline is
#: the client's own machine doing free work.
SERVER_STAGES = ("analyze", "transcribe", "select", "reedit", "effects")


def ticket(
    *,
    stage: str,
    kind: str,
    task: str | None = None,
    kwargs: dict[str, Any] | None = None,
    job_id: str | None = None,
    media_sec: float = 0.0,
    frame_count: int | None = None,
    client_step: dict[str, Any] | None = None,
) -> dict[str, Any]:
    """The ticket for a paid stage that is STARTING.

    ``task``/``kwargs`` let the server re-enqueue the identical worker job with
    no help from the client (nothing is re-uploaded — the frames, proxies or
    WAVs this stage needs are already stored, and only a SUCCESSFUL stage purges
    them). ``client_step`` is the alternative for the one paid stage the client
    drives itself, the synchronous ``plan`` call: it carries the request body to
    repeat, because those numbers (the voiceover's length, the clip durations)
    exist only on the user's machine.
    """
    state: dict[str, Any] = {
        "v": TICKET_VERSION,
        "stage": stage,
        "kind": kind,
        "media_sec": round(max(0.0, float(media_sec or 0.0)), 3),
    }
    if task:
        state["task"] = task
        state["kwargs"] = dict(kwargs or {})
    if job_id:
        state["job_id"] = job_id
    if frame_count is not None:
        state["frame_count"] = int(frame_count)
    if client_step:
        state["client_step"] = client_step
    return state


def mark_paused(
    state: dict[str, Any] | None,
    *,
    stage: str | None = None,
    window: str | None = None,
    resets_at: datetime | None = None,
    wallet_can_cover: bool = False,
    wallet_satang: int = 0,
    run_id: str | None = None,
    message: str | None = None,
    at: datetime | None = None,
) -> dict[str, Any]:
    """The same ticket, stamped with why and when it stopped.

    A ticket that is missing (a pause from a build before this existed, or a
    stage that never wrote one) still produces a usable one: the pause facts
    alone are enough for the client to show the reset time and the top-up
    offer, and the endpoint infers the stage from the project's own artifacts.
    """
    out = dict(state or {})
    out.setdefault("v", TICKET_VERSION)
    if stage and not out.get("stage"):
        out["stage"] = stage
    out["paused"] = True
    out["reason"] = "quota"
    out["paused_at"] = runs.iso(at or datetime.now(UTC))
    out["window"] = window
    out["resets_at"] = runs.iso(resets_at)
    out["wallet_can_cover"] = bool(wallet_can_cover)
    out["wallet_satang"] = int(wallet_satang or 0)
    if run_id:
        out["paused_run_id"] = run_id
    if message:
        out["message"] = message
    return out


def is_current(state: Any) -> bool:
    """A ticket this code understands (and not, say, a hand-edited row)."""
    return isinstance(state, dict) and int(state.get("v") or 0) == TICKET_VERSION


def estimate_for(
    state: dict[str, Any] | None,
    *,
    engine: str | None = None,
    precision: str | None = None,
) -> Estimate | None:
    """What finishing costs: the INTERRUPTED stage only.

    Rebuilt from the ticket's own kind and the footage the server measured when
    that stage started — never from the project's full clip list, which would
    price the boundaries already paid for all over again.
    """
    if not is_current(state):
        return None
    kind = str((state or {}).get("kind") or "")
    if kind not in estimator.MODE_PROFILES:
        return None
    media = float((state or {}).get("media_sec") or 0.0)
    frames = (state or {}).get("frame_count")
    return estimator.estimate_run(
        kind=kind,
        engine=engine,
        precision=precision,
        clip_secs=[media] if media > 0 else (),
        frame_count=int(frames) if frames is not None else None,
    )


async def quota_view(
    session: AsyncSession,
    user: User,
    est: Estimate | None,
    now: datetime | None = None,
) -> dict[str, Any]:
    """Can the account pay for ``est`` right now, and if not, when or with what.

    ``session`` must already be on the ``core`` search_path. The shape mirrors
    ``POST /usage/estimate`` (docs/token-billing-design.md §19) so a client can
    reuse the same reader: ``fits`` is ``plan`` | ``wallet`` | ``none``, ``pct``
    is this run's share of each enforced window, and the money is baht (satang),
    never tokens.
    """
    at = now or datetime.now(UTC)
    if limits_mod.is_unlimited(user):
        return {
            "fits": "plan", "pct": {}, "wallet_satang": 0, "balance_satang": 0,
            "binding": None, "resets_at": None, "unlimited": True,
        }
    account = await get_account(session, int(user.id))
    plan = runs.effective_plan(user, account, at)
    views = runs.enforced_windows(plan, account, at)
    tokens = int(est.tokens) if est is not None else 0
    pct = {
        v.key: (round(tokens / v.limit * 100, 1) if v.limit > 0 else 100.0) for v in views
    }
    tightest = runs.binding_window(views)
    headroom = tightest.headroom if tightest is not None else tokens
    overflow = max(0, tokens - max(0, headroom))
    need_satang = wallet.satang_for_tokens(overflow)
    spare = 0
    if account is not None:
        spare = await wallet.balance(session, int(user.id), at) - int(
            account.wallet_reserved_satang or 0
        )
    fits = "plan"
    if overflow > 0:
        fits = "wallet" if spare >= need_satang else "none"
    return {
        "fits": fits,
        "pct": pct,
        "wallet_satang": need_satang,
        "balance_satang": max(0, spare),
        "binding": tightest.key if tightest is not None else None,
        "resets_at": runs.iso(tightest.resets_at) if tightest is not None else None,
        "unlimited": False,
    }

File: web/src/lib/resume.ts
/**
 * Continuing a project the server parked because the plan's quota ran out.
 *
 * The server is the only place that knows where a run stopped: it records the
 * furthest pipeline BOUNDARY a project completed (`stage`) plus a ticket for
 * the boundary it was attempting, and prices a resume on that one stage. This
 * module is the client half of that contract — parsing, vocabulary and the
 * sentences a person reads. It holds no state and touches no network, so the
 * pipeline, the cards and the tests all read the same definitions.
 *
 * Two rules the whole feature rests on:
 *
 *   1. The ORDER of stages is never hardcoded here. It differs per mode and
 *      arrives as `stages`; this module only knows how to NAME a stage.
 *   2. Money is satang and quota is percent. Nothing turns either back into a
 *      token count, because the server never sends one.
 *
 * Contract: the backend's RESUME CONTRACT (GET/POST /videos/{uid}/resume).
 */

import {
  estimateBlockLine,
  estimateLine,
  formatBaht,
  limitLabel,
  whenBack,
  type LimitKey,
  type UsageEstimate
} from './usageLimits'
import type { ProjectMode, ProjectStep } from './projectFlow'

// ── shapes ───────────────────────────────────────────────────────────────────

export type ResumeRunsOn = 'server' | 'client'
export type ResumeAction = 'server_job' | 'client_step' | 'already_running' | 'nothing_to_resume'
/** `ticket` = priced on the stage the server recorded; `declared` = priced on
 * the lengths declared at creation (older build, display only); `free` = the
 * next step is local and costs nothing. */
export type EstimateSource = 'ticket' | 'declared' | 'free'

/**
 * The live quota answer that rides on every resume payload.
 *
 * Deliberately the RAW server shape (`wallet_satang`, `resets_at`): it is
 * `POST /usage/estimate`'s response plus `balance_satang`, so every sentence
 * `usageLimits` already writes for an estimate works on it unchanged.
 */
export interface ResumeQuota extends UsageEstimate {
  /** Spendable top-up balance right now, in satang. */
  balance_satang: number
}

/** A call the client must repeat itself, because only it has the numbers. */
export interface ResumeClientStep {
  method: string
  path: string
  body: Record<string, unknown>
}

/** `GET /videos/{uid}/resume` — works on any local project, paused or not. */
export interface ResumeState {
  uid: string
  /** The project row's status, e.g. `paused_quota`, `processing`, `done`. */
  status: string
  mode: string
  paused: boolean
  reason: string | null
  /** Furthest COMPLETED boundary; null on a project made before stages existed. */
  stage: string | null
  /** What continuing will do; null means nothing is left. */
  nextStage: string | null
  /** This mode's boundary order, server-supplied. Never hardcode it. */
  stages: string[]
  runsOn: ResumeRunsOn | null
  clientStep: ResumeClientStep | null
  pausedAt: string | null
  /** The window that ran out (frozen at pause time). */
  window: LimitKey | null
  windowResetsAt: string | null
  /** The server's own Thai sentence, only while paused. */
  message: string | null
  resumable: boolean
  /** False ⇒ continuing costs nothing (the next step is local). */
  charges: boolean
  estimateSource: EstimateSource
  quota: ResumeQuota
}

/** `POST /videos/{uid}/resume` — the same keys, plus what the call did. */
export interface ResumeOutcome extends ResumeState {
  action: ResumeAction
  jobId: string | null
  runId: string | null
  resumed: boolean
  detail: string | null
}

// ── parsing ──────────────────────────────────────────────────────────────────

function str(v: unknown): string | null {
  return typeof v === 'string' && v.trim() ? v : null
}

function num(v: unknown, fallback = 0): number {
  return typeof v === 'number' && Number.isFinite(v) ? v : fallback
}

function limitKey(v: unknown): LimitKey | null {
  return v === 'five_hour' || v === 'weekly' || v === 'monthly' ? v : null
}

function pctMap(v: unknown): Partial<Record<LimitKey, number>> {
  if (!v || typeof v !== 'object') return {}
  const out: Partial<Record<LimitKey, number>> = {}
  for (const [k, n] of Object.entries(v as Record<string, unknown>)) {
    const key = limitKey(k)
    if (key && typeof n === 'number' && Number.isFinite(n)) out[key] = n
  }
  return out
}

/**
 * The quota block, defaulted to "the plan covers it".
 *
 * A missing or malformed block must never read as "you cannot continue": the
 * server refuses nothing here — it starts the run and pauses again if the
 * window is still full — so an unreadable estimate should offer the plain
 * "ทำต่อ" rather than push a top-up nobody needs.
 */
export function parseQuota(v: unknown): ResumeQuota {
  const o = (v && typeof v === 'object' ? v : {}) as Record<string, unknown>
  const fits = o.fits === 'wallet' || o.fits === 'none' ? o.fits : 'plan'
  return {
    fits,
    pct: pctMap(o.pct),
    wallet_satang: num(o.wallet_satang),
    balance_satang: num(o.balance_satang),
    binding: limitKey(o.binding),
    resets_at: str(o.resets_at),
    unlimited: o.unlimited === true
  }
}

function parseClientStep(v: unknown): ResumeClientStep | null {
  if (!v || typeof v !== 'object') return null
  const o = v as Record<string, unknown>
  const path = str(o.path)
  if (!path) return null
  return {
    method: str(o.method) ?? 'POST',
    path,
    body: (o.body && typeof o.body === 'object' ? o.body : {}) as Record<string, unknown>
  }
}

function stageList(v: unknown): string[] {
  return Array.isArray(v) ? v.filter((s): s is string => typeof s === 'string' && !!s) : []
}

/** `GET /videos/{uid}/resume`, defensively. Unknown fields are ignored. */
export function parseResumeState(body: unknown): ResumeState {
  const o = (body && typeof body === 'object' ? body : {}) as Record<string, unknown>
  const source = o.estimate_source
  return {
    uid: str(o.uid) ?? '',
    status: str(o.status) ?? '',
    mode: str(o.mode) ?? 'dub_first',
    // `paused` is the server's own word for it; the status is the fallback so
    // an older payload that omits the flag still reads correctly.
    paused: o.paused === true || o.status === 'paused_quota',
    reason: str(o.reason),
    stage: str(o.stage),
    nextStage: str(o.next_stage),
    stages: stageList(o.stages),
    runsOn: o.runs_on === 'server' || o.runs_on === 'client' ? o.runs_on : null,
    clientStep: parseClientStep(o.client_step),
    pausedAt: str(o.paused_at),
    window: limitKey(o.window),
    windowResetsAt: str(o.window_resets_at),
    message: str(o.message),
    resumable: o.resumable === true,
    charges: o.charges === true,
    estimateSource: source === 'ticket' || source === 'declared' ? source : 'free',
    quota: parseQuota(o.quota)
  }
}

/** `POST /videos/{uid}/resume`. All four actions send the same key set, so
 * nothing here may branch on a key being present. */
export function parseResumeOutcome(body: unknown): ResumeOutcome {
  const o = (body && typeof body === 'object' ? body : {}) as Record<string, unknown>
  const action = o.action
  return {
    ...parseResumeState(body),
    action:
      action === 'server_job' ||
      action === 'client_step' ||
      action === 'already_running' ||
      action === 'nothing_to_resume'
        ? action
        : 'nothing_to_resume',
    jobId: str(o.job_id),
    runId: str(o.run_id),
    resumed: o.resumed === true,
    detail: str(o.detail)
  }
}

// ── vocabulary ───────────────────────────────────────────────────────────────

/**
 * Stage name → the step the local pipeline calls it.
 *
 * `render_final` is the one that forks: the dub chain calls it
 * `final_rendering`, the speech chain `rendering` — hence `stageStep` taking
 * the mode. A stage with no local equivalent (`reedit`, `effects`) maps to
 * null: those sit outside the pipeline order and the caller handles them.
 */
const STAGE_STEPS: Record<string, ProjectStep> = {
  imported: 'imported',
  proxy: 'analyzing',
  analyze: 'analyzing',
  render_silent: 'silent_rendering',
  voiceover: 'waiting_vo',
  plan: 'planning',
  extract_audio: 'extracting_audio',
  transcribe: 'transcribing',
  select: 'selecting',
  done: 'done'
}

function isSpeechish(mode: string): boolean {
  return mode === 'talking_head' || mode === 'speech_highlights' || mode === 'speech_scenes'
}

export function stageStep(stage: string | null, mode: string): ProjectStep | null {
  if (!stage) return null
  if (stage === 'render_final') return isSpeechish(mode) ? 'rendering' : 'final_rendering'
  return STAGE_STEPS[stage] ?? null
}

/**
 * What a stage is called to a person — a NOUN phrase ("ถอดเสียง"), not the
 * progressive label the progress bar uses, because these read as "ทำต่อจาก …"
 * and "ขั้นถัดไป: …".
 */
const STAGE_LABELS: Record<string, string> = {
  imported: 'นำเข้าคลิป',
  proxy: 'เตรียมวิดีโอให้ AI',
  analyze: 'ให้ AI วิเคราะห์วิดีโอ',
  render_silent: 'ตัดวิดีโอ (ยังไม่มีเสียงพากย์)',
  voiceover: 'ใส่เสียงพากย์',
  plan: 'วางแผนไทม์ไลน์ตามเสียงพากย์',
  render_final: 'เรนเดอร์วิดีโอสุดท้าย',
  extract_audio: 'แยกเสียงจากคลิป',
  transcribe: 'ถอดเสียง',
  select: 'เลือกช่วงเด่น',
  done: 'เสร็จแล้ว',
  // Side stages — outside the pipeline order, but they can be `next_stage`.
  reedit: 'ให้ AI ตัดใหม่',
  effects: 'เอฟเฟกต์การซูม'
}

export function stageLabel(stage: string | null): string {
  if (!stage) return 'ขั้นตอนถัดไป'
  return STAGE_LABELS[stage] ?? stage
}

/** "ขั้นที่ 3 จาก 8", or null when the stage is not one of this mode's. */
export function stagePosition(state: ResumeState): string | null {
  if (!state.nextStage || state.stages.length === 0) return null
  const i = state.stages.indexOf(state.nextStage)
  if (i < 0) return null
  return `ขั้นที่ ${i + 1} จาก ${state.stages.length}`
}

// ── sentences ────────────────────────────────────────────────────────────────

/**
 * The headline of a paused project. The plan's own word for the window comes
 * first because that is the thing that ran out; the server's sentence is kept
 * for the body, where it can say more than a headline has room for.
 */
export function pausedHeadline(state: ResumeState): string {
  if (!state.paused) return ''
  return state.window ? `หยุดไว้ชั่วคราว — ${limitLabel(state.window)}หมด` : 'หยุดไว้ชั่วคราว'
}

/**
 * When the window that ran out comes back, in the VIEWER's timezone — which
 * the server's own message cannot do. Empty when no time was frozen.
 */
export function windowResetLine(state: ResumeState, now: Date = new Date()): string {
  const back = whenBack(state.windowResetsAt, now)
  if (!back) return ''
  const label = state.window ? limitLabel(state.window) : 'โควตา'
  return `${label}จะรีเซ็ต${back.startsWith('อีก') ? '' : ' '}${back}`
}

/** "ทำต่อจากขั้น: ถอดเสียง" — where continuing picks up. */
export function nextStageLine(state: ResumeState): string {
  if (!state.nextStage) return 'ไม่มีขั้นตอนค้างอยู่'
  const pos = stagePosition(state)
  return `ทำต่อที่: ${stageLabel(state.nextStage)}${pos ? ` · ${pos}` : ''}`
}

/**
 * What continuing would cost, in the same percent-of-window language the
 * wizard uses. Null when the next step is local (nothing is charged) or the
 * account has no enforced windows.
 */
export function resumeCostLine(state: ResumeState): string | null {
  if (!state.charges) return 'ขั้นตอนถัดไปทำบนเครื่องนี้ — ไม่ใช้โควตาเพิ่ม'
  const line = estimateLine(state.quota)
  if (!line) return null
  // A ticket is priced on the stage the server measured; a `declared` estimate
  // was worked out from the lengths given at creation and is a rough guide.
  return state.estimateSource === 'declared' ? `${line} (ประมาณการคร่าว ๆ)` : line
}

/** Why the plan cannot pay for it, and what the balance would do instead. */
export function resumeBlockLine(state: ResumeState, now: Date = new Date()): string | null {
  if (!state.charges) return null
  return estimateBlockLine(state.quota, now)
}

/** True when continuing needs the user's consent to spend the top-up balance. */
export function needsWalletConsent(state: ResumeState): boolean {
  return state.charges && state.quota.fits === 'wallet'
}

/**
 * The label on the continue button.
 *
 * `none` still offers a press: the server does not refuse a resume that cannot
 * fit — it starts the run, the guard stops it and the project parks again — so
 * the honest word is "ลองทำต่อ", not a disabled button that explains nothing.
 */
export function resumeActionLabel(state: ResumeState): string {
  if (!state.charges) return 'ทำต่อ'
  if (state.quota.fits === 'wallet') {
    return `ใช้ยอดเงินคงเหลือ ${formatBaht(state.quota.wallet_satang)} ทำต่อ`
  }
  if (state.quota.fits === 'none') return 'ลองทำต่อ'
  return 'ทำต่อ'
}

/** "ยอดเงินคงเหลือ ฿120.00" — shown beside the wallet action. */
export function balanceLine(state: ResumeState): string {
  return `ยอดเงินคงเหลือ ${formatBaht(state.quota.balance_satang)}`
}

/**
 * The one case that genuinely cannot work, said out loud.
 *
 * A project's footage lives in this browser's own storage. The server holds a
 * synced copy of most of it, so opening the project elsewhere usually works —
 * but when the next boundary runs on the CLIENT and the footage is not
 * reachable from here, continuing is impossible and nothing else in the UI
 * would say why. `footageReachable === null` means "not checked yet" and
 * produces no claim either way.
 */
export function localFootageWarning(
  state: ResumeState,
  footageReachable: boolean | null
): string | null {
  if (state.runsOn !== 'client') return null
  if (footageReachable !== false) return null
  return 'ขั้นตอนถัดไปทำบนเครื่องนี้ แต่ไม่พบไฟล์วิดีโอของโปรเจกต์นี้ในเบราว์เซอร์นี้ — ไฟล์เก็บไว้ต่อเบราว์เซอร์ ทำต่อได้จากเบราว์เซอร์หรืออุปกรณ์ที่สร้างโปรเจกต์นี้เท่านั้น'
}

/** What the POST actually did, for the toast after pressing "ทำต่อ". */
export function outcomeMessage(outcome: ResumeOutcome): string {
  switch (outcome.action) {
    case 'server_job':
      return `ทำต่อแล้ว — ${stageLabel(outcome.nextStage)}`
    case 'already_running':
      return 'งานนี้กำลังทำอยู่แล้ว'
    case 'client_step':
      return `ทำต่อแล้ว — ${stageLabel(outcome.nextStage)}`
    case 'nothing_to_resume':
      return outcome.detail ?? 'ไม่มีขั้นตอนค้างอยู่'
  }
}

/**
 * Where a resumed project should sit locally once the pause is cleared.
 *
 * Driven by the SERVER's status, not by whatever the browser last wrote: a
 * project paused on one machine and resumed on another has no useful local
 * step to fall back to. `fallback` covers `pending`, which only says "nothing
 * is running".
 */
export function stepAfterResume(
  state: ResumeState,
  mode: ProjectMode,
  fallback: ProjectStep
): ProjectStep {
  if (state.status === 'done') return 'done'
  if (state.status === 'waiting_vo') return 'waiting_vo'
  if (state.status === 'processing') return stageStep(state.nextStage, mode) ?? fallback
  return fallback
}
